### PR TITLE
Warn on unresolved variables in swarm.json

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Fixes https://github.com/giantswarm/cli/issues/21

Any variables defined in swarm.json that have not been supplied will not be resolved, which may cause issues further down, so this change introduces a check that runs after creating the service definition, searches for any tokens that look like unresolved variables, and prints a warning if it finds any.

So, with the helloworld example, if the user doesn't use
```
swarm up --var=domain=helloworld-$USER.gigantic.io
```
and just does
```
swarm up
```
then the user will be warned:

```
Josephs-MacBook-Pro:helloworld joseph$ ../cli/swarm up
We found a token in swarm.json that looks like a variable that has not been resolved:
- $domain
Please make sure to define any variables that are required.
Creating 'helloworld' in the 'salisbury_joe/testing' environment…
Cannot create service due to a validation error in 'swarm.json'.

    Validation error: invalid service definition: invalid format for domains: Port must be a number, got '$domain': strconv.ParseInt: parsing "$domain": invalid syntax

### CODE REVIEW

**Comment by marians** [comment]
Will test...

**Comment by ewoutp** [comment]
Please add some `int-tests` for this.

**Comment by zyndiecate** [comment]
IMO this is way too much text. Maybe it is a personal thing and I know people have well defined opinions on that.

File: create.go [hunk 1]
```diff
@@ -91,6 +92,26 @@ func processCreate(defFilePath, varFile string, variables StringSlice, name stri
 		Quitf(defaultErrorMsgTemplate, err)
```
Comment (zyndiecate) [inline create.go]: Unfortunately I cannot comment on the line I would like to. In line 90 there is the call to `create.NewDefinition`. This is basically used when we handle any kind of definitions. E.g. the update command is using it as well. So I would like to put this validation stuff maybe inside this method, so we can also use it in other commands then `create`.

**Comment by marians** [comment]
I'm all for this amount of text. Don't know how we could make it any shorter.

In terms of writing style, we consciously decided use natural language instead of abbreviated technical "Error: unparsed variable found", which you would then have to look up somewhere to really understand.

I'd rather show a bit more of a warning than let users run into a troubleshooting situation that costs their or our time.

**Comment by zyndiecate** [comment]
> I'm all for this amount of text. Don't know how we could make it any shorter.

I know. For me this does not help. I am a developer and I am annoyed by those amounts of information for essentially "simple" things. I know it is not simple what we do and what we need to cover, but IMO devs only are annoyed by the need to deal with such a mess of information. We show prose as we are writing a novel. I want to see short and clear information. I also miss some newlines here. I just see a block of text and I am just really disappointed to figure out what the damn tool want to say.

**Comment by marians** [comment]
> I also miss some newlines here.

Here we agree. :)

**Comment by JosephSalisbury** [comment]
with a newline, so it's easy to scan:

```
Josephs-MacBook-Pro:Repos joseph$ ./cli/swarm up ./helloworld/swarm.json
We found a token in ./helloworld/swarm.json that looks like a variable that has not been resolved:
- $domain
Please make sure to define any variables that are required.

Creating 'helloworld' in the 'salisbury_joe/testing' environment…
Cannot create service due to a validation error in './helloworld/swarm.json'.

    Validation error: invalid service definition: invalid format for domains: Port must be a number, got '$domain': strconv.ParseInt: parsing "$domain": invalid syntax
```

File: int-tests/104-unresolved-variables.t [hunk 2]
```diff
@@ -0,0 +1,29 @@
+Unresolved variables:
```
Comment (ewoutp) [inline int-tests/104-unresolved-variables.t]: I would prefer to have this test in the 0..99 range so we always run it.
Everything above 100 is only run when doing `make run-ful-int-tests` which is so slow that we forget to run it.

File: int-tests/104-unresolved-variables.t [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  $ if [ -z ${ALL} ]; then exit 80; fi # Skip unless ALL is non-empty
```
Comment (ewoutp) [inline int-tests/104-unresolved-variables.t]: Please remove this so we always run it

**Comment by marians** [comment]
How about this:

```
$ swarm create swarm1.json 
We found tokens in swarm1.json that look like variables that have not been resolved:

    $orgname
    $image

Please make sure to define any variables that are required.

Creating 'redis-example' in the 'marian/dev' environment…
Cannot create service due to a validation error in 'swarm1.json'.

    Validation error: invalid service definition: Invalid namespace part: $orgname: Not a valid docker image. Format: [<registry>/]<repository>[:<version>]

```

**Comment by ewoutp** [comment]
Just wondering what happens if I run something like this:

```
{
  "service": {
     "image": "foo:1.2.3",
     "args": ["$MY_ENV"],
     "env": {
         "MY_ENV":"something"
      }
  }
}
```

**Comment by JosephSalisbury** [comment]
@ewoutp this will still raise the warning - the warning is largely best case until we get a fancier variable sustitution

**Comment by ewoutp** [comment]
Can't we detect this case. To me it sounds a bit annoying if I would get warnings for cases like this.

**Comment by marians** [comment]
@ewoutp I don't think that's a very common case. We grepped through all service definitions and didn't find any `$` signs in there, except for those which were clearly erroneously unreplaces variables.

**Comment by JosephSalisbury** [comment]
RFR plz

**Comment by zyndiecate** [comment]
IIWFYLGTM

**Comment by marians** [comment]
> IIWFYLGTM

@JosephSalisbury This means "If it works for you, looks good to me" :)

**Pull request merged**